### DISCUSSION TITLE
Containerising the project

### DISCUSSION THREAD

**Discussion by vigneshwar221B** [developersIndia/developersIndia#61] [Ideas]
I think it's good to containerise the project from the start and let the developers work on top of it. It will also be useful and hassle-free for new contributors. Instead of telling them to install node and related libraries, we can tell them to just install docker. Once they have docker installed, they can run/stop the project with simple one liners like `docker-compose up/down`. I think it provides a nice abstraction layer and as we can also have live reloading, this transition should be seamless. 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by jarmos-san** [comment]
Continuing the discussion we had on Discord;

@Bhupesh-V approves dockerising the project as well. Could you provide few more details like:

1. An example `Dockerfile` (or a `docker-compose.yml`) file(s).
2. Explanation of how to set it up because we'll need to update the "[contributing guidelines](https://github.com/developersIndia/website/blob/master/.github/CONTRIBUTING.md)" as well.
3. Are you willing to take up this responsibilities? If so, please let us know about your plans. Sharing a simple `Dockerfile` (or a` docker-compose.yml`) isn't enough, the project requires updating the docs where necessary as well.

**Comment by vigneshwar221B** [comment]
1. Example `Dockerfile` & `docker-compose.yml` files:

    `Dockerfile`:

    ```dockerfile
    # Using node alpine as base image
    FROM node:alpine
    
    # Create the directory on the node image 
    # where our Next.js app will live
    RUN mkdir -p /app
    
    # Set /app as the working directory
    WORKDIR /app
    
    # Copy package.json and package-lock.json
    # to the /app working directory
    COPY package*.json /app
    
    # Install dependencies in /app
    RUN npm install
    
    # Copy the rest of our Next.js folder into /app
    COPY . /app
    
    # Ensure port 3000 is accessible to our system
    EXPOSE 3000
    
    # Run npm dev, as we would via the command line 
    CMD ["npm", "run", "dev"]
    ```

    `docker-compose.yml`:

    ```yaml
    version: "3"
    
    services:
      web:
        build:
          context: .
          dockerfile: Dockerfile
        container_name: dev-india
        restart: always
        volumes:
          - ./:/app
          - /app/node_modules
          - /app/.next
        ports:
          - 3000:3000
    ```
    I've also made some changes in `next.config.js` to enable hot reloading which you will see in my upcoming PR.

2. In the `README.md`, we have to change the **prerequisites and installation** sections.  As we are using [Docker](https://www.docker.com), the prerequisites are gonna be having just `git` and `docker` installed. And as for **installation** section, they just need to run the command `docker-compose up -d --build` to run the server and `docker-compose down` to stop the server. 

3. Yes, I'm willing to take the responsibility. I've made this changes only considering what your current plans are. You've mentioned you're only planning on deploying this site in [Vercel](https://vercel.com) as a static website. So, this current configuration should work out of the box. I've seen a discussion here suggesting [Nginx](https://www.nginx.com) as a reverse proxy. When we make that change, we have to write few more lines in `docker-compose.yml` but that should be it. And this applies to any new additions to the project including having a DB, redis, etc.

**Reply by jarmos-san** [reply]
Wonderful @vigneshwar221B thanks for sharing the details with us! :heart: We'll use this thread as a source for all future discussions.

As for your PR about hot-reloading, please open an issue thread detailing what it does, why it's needed, what changes to source code will need to be introduced & a link to the documentations for ours & other contributor's reference. And don't forget to reference that issue thread in your PR. Also include documentation changes (if applicable) where necessary.

**NOTE**: Next.js supports Fast Refresh out-of-the-box, see "[Basic Features: Fast Refresh](https://nextjs.org/docs/basic-features/fast-refresh)" for reference.

**Reply by tarunkishore2303** [reply]
I would love to work with docker here. Ping me if you need any kind of help !
